### PR TITLE
Adds JSON Exporter

### PR DESCRIPTION
**Pull request by athityakumar**

### CODE REVIEW

File: lib/daru/io/exporters/json.rb [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+        #   #     }
+        #   #   },
+        #   #   {
+        #   #     "Lyanna Stark": {
+        #   #     "gender": "Female",
+        #   #     "age": 36
+        #   #     }
+        #   #   }
+        #   # ]
+        def initialize(dataframe, path, orient: :index, pretty: false, **jsonpaths)
```
Comment (athityakumar) [inline lib/daru/io/exporters/json.rb]: The `:orient` option is inspired from [pandas](https://github.com/pandas-dev/pandas/blob/v0.20.3/pandas/core/generic.py#L1129-L1245), which can be set to `:records`, `:split`, `:values` and defaults to `:index`.

File: lib/daru/io/exporters/json.rb [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+        #   # child   Jon Snow         18       Male
+        #   # dad   Rhaegar Ta         54       Male
+        #   # mom   Lyanna Sta         36     Female
+        #
+        #   Daru::IO::Exporters::JSON.new(
+        #     df,
+        #     'filename.json',
+        #     orient: :records,
+        #     pretty: true,
+        #     name: '$..specific..name',
```
Comment (zverok) [inline lib/daru/io/exporters/json.rb]: Why `..` here? In JsonPath search, `..` means "then skip any number of levels". For creation, I believe, `.` is appropriate and `..` should be treated as undefined behavior.

File: lib/daru/io/exporters/json.rb (hunk 1 above)
Comment (zverok) [inline lib/daru/io/exporters/json.rb]: I don't believe it does any good. Can we at least for ver.0 try to do _everyting_ with JsonPath? Can all `orient` options be clearly expressed with JsonPath?..

File: lib/daru/io/exporters/json.rb [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+        #   #   },
+        #   #   {
+        #   #     "Lyanna Stark": {
+        #   #     "gender": "Female",
+        #   #     "age": 36
+        #   #     }
+        #   #   }
+        #   # ]
+        def initialize(dataframe, path, orient: :index, pretty: false, **jsonpaths)
+          optional_gem 'json'
```
Comment (zverok) [inline lib/daru/io/exporters/json.rb]: Isn't json part of stdlib?..

File: lib/daru/io/exporters/json.rb [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+          @json_content    = process_json_content
+
+          File.open(@path, 'w') { |file| file.write(process_json_string) }
+        end
+
+        private
+
+        def process_json_string
+          return ::JSON.pretty_generate(@json_content) if @pretty
+          ::JSON.generate(@json_content)
```
Comment (zverok) [inline lib/daru/io/exporters/json.rb]: As I've said in another PR, I believe you are too inclined to the pattern sometimes. For me, this version is cleaner:
```ruby
@pretty ? ::JSON.pretty_generate : ::JSON.generate
```
If you really want to, you can "hack" through Ruby code and even do something like:
```ruby
::JSON.generate(@json_content, @pretty ? JSON::PRETTY_STATE_PROTOTYPE : {})
```
...but, honestly, I am not sure it is better :)

File: lib/daru/io/exporters/json.rb (hunk 1 above)
Comment (athityakumar) [inline lib/daru/io/exporters/json.rb]: But isn't this exactly what `:orient` is supposed to take care of? I mean, JsonPath is supposed to only create nested Hashes? Maybe a block to manipulate the finally received `@json_content` would be better?

**Comment by athityakumar** [comment]
@zverok - I've removed the `:orient` option as per the review. But this now means that `to_json` will always return something like an Array of (nested) Hashes, rather than other (manipulation) options like Hash of Hashes or Array of Arrays. :-/

File: lib/daru/io/exporters/json.rb [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        end
+
+        def call
+          @jsonpath_hash ||= @dataframe.vectors.to_a.map { |v| {v => "$.#{v}"} }.reduce(:merge)
+          @vectors         = @jsonpath_hash.keys
+          @jsonpaths       = process_jsonpath
+          @json_content    = @dataframe.map_rows_with_index do |row, idx|
+            init_hash(@jsonpaths, @vectors, row, idx)
+          end
+
```
Comment (athityakumar) [inline lib/daru/io/exporters/json.rb]: After having removed the `:orient` option, JsonPath doesn't seem too flexible for various formats such as Array of Array, Hash of Hash, etc. as it always returns an Array of Nested Hash. But I think that allowing users to manipulate this formed `@json_content` via a block, would provide greater flexibility. WDYT?

Ping @zverok @v0dro @lokeshh

Comment (zverok) [inline lib/daru/io/exporters/json.rb]: As I've written in private, let's create GH issue with 5-7 examples of reasonable target outputs of the same dataframe, and brainstorm the minimal flexible API?..

Comment (athityakumar) [inline lib/daru/io/exporters/json.rb]: I've added some more (intended) use-cases on issue #10. Please have a look. 😄

Comment (zverok) [inline lib/daru/io/exporters/json.rb]: :+1: Looking at it

Comment (athityakumar) [inline lib/daru/io/exporters/json.rb]: I've dded the `:orient` & block manipulation, as per discussion on issue #10. Please review when you're free. :smile:

**Comment by v0dro** [comment]
Whats the update on this issue?

**Comment by athityakumar** [comment]
@v0dro - I've added `:orient` and block manipulation, as per discussion on issue #10. Requires a final review.

File: lib/daru/io/exporters/json.rb [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+            (JsonPath.new(x).path - %w[$ ${ . .. ..{]).map do |y|
+              v = y.delete("'.[]{")
+              next v.to_i if v.to_i.to_s == v
+              v.to_sym
+            end
+          end
+        end
+
+        def validate_params
+          return if %i[index records split values].include?(@orient)
```
Comment (zverok) [inline lib/daru/io/exporters/json.rb]: Constant?

File: lib/daru/io/exporters/json.rb [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+            init_hash_rec(jsonpaths[1..-1], hash[key], jsonpath_key, row, idx)
+          end
+          hash
+        end
+
+        def init_hash(jsonpaths, jsonpath_keys, row, idx)
+          first, *rest = jsonpaths.map.with_index do |path, i|
+            init_hash_rec(path, Hash.new { |h,k| h[k] = Hash.new(&h.default_proc) }, jsonpath_keys[i], row, idx)
+          end
+          rest.each { |r| first = deep_merge(first, r) }
```
Comment (zverok) [inline lib/daru/io/exporters/json.rb]: reduce

**Pull request merged**